### ISSUE TITLE
migrate away from vdm

### ISSUE THREAD

**Issue by wardi** [ckan/ideas#41]
IIUC vdm is holding us back from upgrading to a more recent SQLAlchemy. Its revision tables also denormalize our data and make writing efficient queries difficult.

Is anyone familiar with alternate revisions-in-the-database projects? Can we integrate revisions into activities somehow?

**Comment by rossjones** [comment]
Just noticed this today: http://docs.sqlalchemy.org/en/rel_1_0/orm/examples.html#versioning-objects

Seem small and simple enough to replace VDM? We can keep the functionality, nobody should be scared to look at the code.

**Comment by rgrp** [comment]
@rossjones that's been in there for ages. You could definitely replace with this - or just rip versioning out entirely ;-)

Do note that the sqlalchemy object versioning method is actually used in the vdm changeset branch i did a few years ago: https://github.com/okfn/vdm/tree/changeset

**Comment by rossjones** [comment]
I think there was some resistance to removing all versioning, as there is currently data lost when just using activities - at least until activities have an associated diff ;)

**Comment by rgrp** [comment]
@rossjones yes you are right :-) I think you could definitely move to simple sqla versioning though what you lose is m2m versioning done right. You may want to look at hte changeset branch in vdm as a) it was for a newer version of sqlalchemy b) it was quite a lot cleaner. That said i think it may have had a different model that means upgrading could be annoying.

**Comment by davidread** [comment]
CKAN used to use the m2m versioning for tags, but this was taken out in 2012: https://github.com/ckan/ckan/commit/496452cb43590f7b9be6a4fb6135a5d226dc4391

We didn't get into 'reverting' revisions, so I think m2m revisioning was dropped for being complicated. But anyway, that's one less thing to worry about here.

So we just need basic 'copy on write' functionality, with a timestamp and the user. A cut-down vdm would do it - we should definitely look at Rufus' branch. The sqlalchemy code looks reasonably similar:
http://docs.sqlalchemy.org/en/latest/_modules/examples/versioned_history/history_meta.html
and I get the impression from a lack of google hits for it that there is little community for it either, so maybe the pain of changing the model to it's one wouldn't be worth it. But maybe we can steal some of the code / understanding into our hacked-down vdm.

**Comment by wardi** [comment]
Great to see huge progress on this by @TkTech https://github.com/ckan/ckan/pull/3485

**Comment by davidread** [comment]
Afaict, to get rid of revisions and vdm, we want this to be complete: https://github.com/ckan/ckan/pull/3972 which saves the full dataset dict in the activity stream. This is not too far off being complete now.
